cpp/open3d/core/Tensor.cpp

```diff
@@ -712,8 +738,58 @@ Tensor Tensor::IndexGet(const std::vector<Tensor>& index_tensors) const {
 
 void Tensor::IndexSet(const std::vector<Tensor>& index_tensors,
                       const Tensor& src_tensor) {
+    if (NumDims() == 0) {
+        const std::string error_prefix =
+                "A 0-D tensor can only be indexed by a 0-D boolean tensor";
+        if (index_tensors.size() != 1) {
+            utility::LogError("{}, but got {} index tensors.", error_prefix,
+                              index_tensors.size());
+        }
+        Tensor index_tensor = index_tensors[0];
+        index_tensor.AssertShape(
+                {}, fmt::format("{}, but got shape {}.",
+                                index_tensor.GetShape().ToString()));
+        index_tensor.AssertDtype(
+                Dtype::Bool, fmt::format("{}, but got dtype {}.",
+                                         index_tensor.GetDtype().ToString()));
+
+        // E.g.
+        // a = np.array(5)
+        // a[np.array(True)]  = 10                 // Works, assigned
+        // a[np.array(True)]  = np.array(10)       // Works, assigned
+        // a[np.array(True)]  = np.array([10])     // Works, assigned
+        // a[np.array(True)]  = np.array([[10]])   // Cannot assign 2D
+        // a[np.array(True)]  = np.array([10, 11]) // Cannot assign 1+ values
```

Comment:
Add unit test for this case. Previously the shape of `src_tensor` it is not checked.